main/src/library/Set.flix

```diff
@@ -264,46 +256,43 @@ namespace Set {
     /// The function `f` must be pure.
     ///
     pub def filter(f: a -> Bool, xs: Set[a]): Set[a] =
-        let Set(s) = xs;
-            Set(List.filter(f, s))
+        foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs)
 
     ///
     /// Returns the result of applying `f` to every element in `xs`.
     ///
     /// Note: The returned set may be smaller than `xs`.
     ///
-    pub def map(f: a -> b & e, xs: Set[a]): Set[b] & e = match xs {
-        case Set(Nil) => Set(Nil)
-        case Set(x :: rs) => insert(f(x), map(f, Set(rs)))
-    }
+    pub def map(f: a -> b & e, xs: Set[a]): Set[b] & e =
+        foldLeft((acc, x) -> insert(f(x), acc), empty(), xs)
 
     ///
     /// Returns the result of applying `f` to every element in `xs` and taking the union.
     ///
-    pub def flatMap(f: a -> Set[b] & e, xs: Set[a]): Set[b] & e = match xs {
-        case Set(Nil) => Set(Nil)
-        case Set(x :: rs) => union(f(x), flatMap(f, Set(rs)))
-    }
+    pub def flatMap(f: a -> Set[b] & e, xs: Set[a]): Set[b] & e =
+        foldLeft((acc, x) -> union(acc, f(x)), empty(), xs)
 
     ///
     /// Replaces the element `x` with `y` if `x` is in `xs`. Otherwise, returns `xs`.
     ///
     /// Note: The returned set may be smaller than `xs`.
     ///
-    pub def replace(x: a, y: a, xs: Set[a]): Set[a] = map(e -> if (e == x) y else e, xs)
+    pub def replace(x: a, y: a, xs: Set[a]): Set[a] =
+        if (memberOf(x, xs)) insert(y, delete(x, xs)) else xs
 
     ///
     /// Returns a pair of sets `(ys, zs)`.
-    //
+    ///
     /// `ys` contains all elements of `xs` that satisfy the predicate `f`.
     /// `zs` contains all elements of `xs` that do not satisfy the predicate `f`.
     ///
     /// The function `f` must be pure.
     ///
     pub def partition(f: a -> Bool, xs: Set[a]): (Set[a], Set[a]) =
-        let Set(s) = xs;
-        let (ys, zs) = List.partition(f, s);
-            (Set(ys), Set(zs))
+        foldLeft((acc, x) ->
+            let (a, b) = acc;
+            if (f(x)) (insert(x, a), b)
+            else (a, insert(x, b)), (empty(), empty()), xs)
```

Comment:
reformat more nicely.

Reply:
Do you prefer it one a single line, or?
I could inline everything, but I would like to avoid 4 calls to fst/snd.

Reply:
No, I want more lines:

```
if (e)
   dostuff
else 
  dostuff
```